ostd/src/mm/page_table/node/entry.rs

```diff
@@ -0,0 +1,159 @@
+        }
+
+        self.pte.set_prop(new_prop);
+
+        // SAFETY:
+        //  1. The index is within the bounds.
+        //  2. We replace the PTE with a new one, which differs only in
+        //     `PageProperty`, so it is still compatible with the current
+        //     page table node.
+        unsafe { self.node.write_pte(self.idx, self.pte) };
+    }
+
+    /// Replaces the entry with a new child.
+    ///
+    /// The old child is returned.
+    ///
+    /// # Panics
+    ///
+    /// The method panics if the given child is not compatible with the node.
+    /// The compatibility is specified by the [`Child::is_compatible`].
+    pub(in crate::mm) fn replace(self, new_child: Child<E, C>) -> Child<E, C> {
+        assert!(new_child.is_compatible(self.node.level(), self.node.is_tracked()));
+
+        // SAFETY: The entry structure represents an existent entry with the
+        // right node information. The old PTE is overwritten by the new child
+        // so that it is not used anymore.
+        let old_child =
+            unsafe { Child::from_pte(self.pte, self.node.level(), self.node.is_tracked()) };
+
+        if old_child.is_none() && !new_child.is_none() {
+            *self.node.nr_children_mut() += 1;
+        } else if !old_child.is_none() && new_child.is_none() {
+            *self.node.nr_children_mut() -= 1;
+        }
+
+        // SAFETY:
+        //  1. The index is within the bounds.
+        //  2. The new PTE is compatible with the page table node, as asserted above.
+        unsafe { self.node.write_pte(self.idx, new_child.into_pte()) };
+
+        old_child
+    }
+
+    /// Splits the entry to smaller pages if it maps to a untracked huge page.
+    ///
+    /// If the entry does map to a untracked huge page, it is split into smaller
+    /// pages mapped by a child page table node. The new child page table node
+    /// is returned.
+    ///
+    /// If the entry does not map to a untracked huge page, the method returns
+    /// `None`.
+    pub(in crate::mm) fn split_if_untracked_huge(self) -> Option<PageTableNode<E, C>> {
+        let level = self.node.level();
+
+        if !(self.pte.is_last(level)
+            && level > 1
+            && self.node.is_tracked() == MapTrackingStatus::Untracked)
+        {
+            return None;
+        }
+
+        let pa = self.pte.paddr();
+        let prop = self.pte.prop();
+
+        let mut new_page = PageTableNode::<E, C>::alloc(level - 1, MapTrackingStatus::Untracked);
+        for i in 0..nr_subpage_per_huge::<C>() {
+            let small_pa = pa + i * page_size::<C>(level - 1);
+            let _ = new_page
+                .entry(i)
+                .replace(Child::Untracked(small_pa, level - 1, prop));
+        }
+
+        let _ = self.replace(Child::PageTable(new_page.clone_raw()));
+
+        Some(new_page)
+    }
+
+    /// Create a new entry at the node.
+    pub(super) fn new_at(node: &'a mut PageTableNode<E, C>, idx: usize) -> Self {
```

Comment:
As discussed earlier, let's mark `new_at` as unsafe.

The following `debug_assert!` can be replaced by safety conditions (keeping the `debug_assert!` is generally okay, but in this particular case it seems redundant since the only caller is `entry` and it does the `assert!`).